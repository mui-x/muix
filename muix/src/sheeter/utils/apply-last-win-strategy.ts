import { TAtomize } from 'reactxx-typings'
import { globalOptions } from '../global-state'

export interface ApplyLastWinStrategyResult {
    style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
    defferedIdxs?: number[]
}
export const enum AttemptType {
    first,
    firstIgnore,
    second
}

export type ApplyLastWinStrategyLow = (values: TAtomize.AtomicWeb[], attemptType: AttemptType) => ApplyLastWinStrategyResult

export const applyLastwinsStrategyRoot = (values: TAtomize.AtomicArray, applyLastWinStrategyLow: ApplyLastWinStrategyLow) => {
    const processDeffereds = globalOptions.processDeffereds
    let res = applyLastWinStrategyLow(values, processDeffereds ? AttemptType.first : AttemptType.firstIgnore)
    if (res.style) return res.style
    processDeffereds(values, res.defferedIdxs, values.state)
    res = applyLastWinStrategyLow(values, AttemptType.second)
    return res.style
}
