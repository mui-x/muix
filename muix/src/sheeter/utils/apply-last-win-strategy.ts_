import { TAtomize } from 'reactxx-typings'

// export interface ApplyLastWinStrategyResult {
//     style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
//     defferedIdxs?: number[]
// }
// export const enum AttemptType {
//     first,
//     firstIgnore,
//     second
// }

