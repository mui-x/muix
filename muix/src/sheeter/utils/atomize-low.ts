import { wrapPseudoPrefixes } from './wrap-pseudo-prefixes'
import { TTyped, TEngine } from 'reactxx-typings'

// platform dependent 
import { platform } from 'reactxx-sheeter'

//*******************************************************
//        adjustAtomizedLow

// linearize and atomize "ruleset"
// result is added to 'atomizedVariants'
export const adjustAtomizedLow = (
    ruleset: TEngine.Rulesets,
    atomizedVariants: TEngine.QueryableItems, path: string,
    pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    processTree(ruleset, atomizedVariants, path, pseudoPrefixes, conditions)
}

export function isToAtomizeArray(obj): obj is TEngine.Ruleset[] { return obj && !(obj as TEngine.Queryables).$r$ && Array.isArray(obj) }
export function isAtomized(obj): obj is TEngine.Queryables { return !obj || (obj as TEngine.Queryables).$r$ }
export function isDeferred(obj): obj is TEngine.Deferred { return obj && (obj as TEngine.Deferred).$d$ }
export function isTemporary(obj): obj is TEngine.TempProc { return obj && (obj as TEngine.TempProc).$t$ }
export function isToAtomize(obj): obj is TEngine.ToAtomize { return obj && !(obj as TEngine.Queryables).$r$ && !(obj as TEngine.Deferred).$d$ && !(obj as TEngine.TempProc).$t$ }

//*******************************************************
//        makeTemporary
export const makeTemporary = (proc: TEngine.TempProc) => {
    proc.$t$ = true
    return proc as any
}

//*******************************************************
//        processTree
export const processTree = (
    value: TEngine.Ruleset | TEngine.Ruleset[], atomizedVariants: TEngine.QueryableItems,
    path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    const items = isToAtomizeArray(value) ? value : [value]
    const indexToPath = (idx: number) => items.length <= 1 ? '' : `[${idx}]`
    items.forEach((it, idx) => {
        if (!it) return
        if (isTemporary(it)) {
            it(atomizedVariants, path + indexToPath(idx), pseudoPrefixes, conditions)
        } else if (isDeferred(it)) {
            atomizedVariants.push(it)
        } else if (isAtomized(it)) {
            if (pseudoPrefixes && pseudoPrefixes.length > 0)
                throw 'Web pseudo properties cannot contain already atomized value'
            it.forEach(it => {
                if (!conditions || conditions.length === 0) //... and no conditions => done
                    pushToAtomizedVariants(atomizedVariants, it, null)
                else { // ... and conditions => merge conditions
                    pushToAtomizedVariants(atomizedVariants, [...it],
                        it.conditions ? [...conditions, ...it.conditions] : conditions)
                }
            })
        } else { // it: toAtomize
            processPseudosAndAtomize(
                it,
                atomizedVariants, path + indexToPath(idx),
                pseudoPrefixes, conditions
            )
        }
    })
}

//*********************************************************
//  PRIVATE
//*********************************************************

//*******************************************************
//        processAllInnerVariants

// recursivelly process web pseudos (:hover, :active) and its queryable parts

const processPseudosAndAtomize = (
    ruleset: TEngine.ToAtomize,
    atomizedVariants: TEngine.QueryableItems, path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions,
) => {

    const inner: TEngine.QueryableItems = []

    // process and remove conditional parts of self and web pseudo (e.g. :hover etc.)
    // push them to 'inner'
    for (const p in ruleset) {
        const value = ruleset[p] as TEngine.ToAtomize
        if (!value || p.charAt(0) === '$' || (typeof value !== 'object' && !isTemporary(value))) continue
        processTree(value, inner, `${path}/${p}`, [...pseudoPrefixes, p], conditions)
        delete ruleset[p]
    }

    // atomize pure ruleset tree with pseudo (and without all conditions, deffer's etc.),
    // then push to result
    atomizeNonVariant(atomizedVariants, wrapPseudoPrefixes(ruleset, pseudoPrefixes), conditions, path)

    // push 'inner' to result
    if (inner.length > 0)
        Array.prototype.push.apply(atomizedVariants, inner)

}

// *********** utils

const pushToAtomizedVariants = (
    atomizedVariants: TEngine.QueryableItems, 
    variant: TEngine.Queryable, conditions: TEngine.Conditions
    ) => {
    if (!variant) return
    if (conditions && conditions.length > 0) variant.conditions = conditions
    atomizedVariants.push(variant)
}

const atomizeNonVariant = (
    atomizedVariants: TEngine.QueryableItems, sourceRuleset: TEngine.Rulesets,
    conditions: TEngine.Conditions, path: string
) => {
    if (!sourceRuleset) return
    if (isDeferred(sourceRuleset)) {
        atomizedVariants.push(sourceRuleset)
        return
    }
    // platform specific atomize: e.g. for WEB convert ruleset to list of atomized CSS classes
    const variant = platform.toPlatformAtomizeRuleset(sourceRuleset, path)
    pushToAtomizedVariants(atomizedVariants, variant, conditions)
}

