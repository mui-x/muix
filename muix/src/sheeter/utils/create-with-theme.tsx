import warning from 'warning'

export const createWithTheme = (valueOrCreator: ((theme) => any) | any, theme) => {
  if (typeof valueOrCreator === 'function') {
    warning(theme, 'Theme expected (ThemeProvider or getDefaultTheme missing)')
    // apply creator to theme:
    return valueOrCreator(theme)
  }
  else
    // return value
    return valueOrCreator
};