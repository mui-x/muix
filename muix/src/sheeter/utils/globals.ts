import { TExtensions, TEngine } from 'reactxx-typings';
import { WidthStore } from '../queryable/$widths/store';
import { toEngineClassName } from 'reactxx-use-sheeter';

export interface Platform extends TExtensions.Platform {
    toPlatformAtomizeRuleset?: TEngine.ToPlatformAtomizeRuleset
    applyLastwinsStrategy?: TEngine.ApplyLastwinsStrategy
    finalizeClassName?: (values: TEngine.AtomicLow) => TEngine.AtomicFinal
    getDefaultTheme?: () => any
}

export const resetPlatform = () => {
    for (const p in platform) delete platform[p]
}

export const assignPlatform = (gl: Platform) => Object.assign(platform, gl)

export const platform: Platform = {}

export const initGlobals = (force: boolean, platformDependentInit: () => void) => {
    if (force) resetPlatform()
    if (platform._sheeter) return

    platform._sheeter = {}
    // fill e.g. actWidth
    platformDependentInit()
    // init store with act width
    platform._sheeter.widthsStore = new WidthStore()

}