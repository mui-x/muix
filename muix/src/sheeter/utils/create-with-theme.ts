import warning from 'warning'
import { isTemporary } from './atomize-low'

export const createWithTheme = (valueOrCreator: ((theme) => any) | any, theme) => {
  if (typeof valueOrCreator === 'function' && !isTemporary(valueOrCreator))
    // apply creator to theme:
    return valueOrCreator(theme)
  else // return value
    return valueOrCreator
}