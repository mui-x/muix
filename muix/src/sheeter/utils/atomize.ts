import { TEngine } from 'reactxx-typings';
import { adjustAtomizedLow, isDeferred } from './atomize-low';
import { createWithTheme } from './create-with-theme';



// muttable
export const atomizeSheet = (sheet: TEngine.SheetOrCreator, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    const sh: TEngine.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const at = atomizeRuleset(sh[p], theme, (path ? path + '.' : '') + p)
        if (at) sh[p] = at
        else delete sh[p]
    }
    return sh
}

// muttable
export const atomizeRuleset = (
    ruleset: TEngine.RulesetOrCreator, theme?, path: string = ''
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme)

    if (!rs) return null

    const list = wrapRuleset([])
    adjustAtomizedLow(rs, list, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapRuleset = ruleset => {
    (ruleset as TEngine.Queryables).$r$ = true
    if (window.__TRACE__)
        ruleset['toJSON'] = toJSON.bind(ruleset)
    return ruleset as TEngine.Queryables
}
function toJSON() {
    return (this as TEngine.Queryables).map(v => {
        if (isDeferred(v)) return 'DEFFERED'
        return v
    })
}

// muttable (at least for native)
export const atomizeStyle = (style: TEngine.Style, theme?) => {
    if (!style) return null

    style = createWithTheme(style, theme)

    const { $web, $native } = style

    if ($web || $native) style = { ...style }

    if (window.isWeb && $web)
        Object.assign(style, $web)
    else if (!window.isWeb && $native)
        Object.assign(style, $native)

    delete style.$web
    delete style.$native
    
    return style
}

