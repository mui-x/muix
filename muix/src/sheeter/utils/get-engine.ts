import { TTyped } from 'reactxx-typings';
import { $width, $web, $native, $hot, $if, $ifelse } from '../queryable/index'
import { toClassNamesWithQuery } from '../utils/to-classnames'
import { mergeRulesets, mergeSheets } from '../utils/merge'
import { atomizeRuleset, atomizeSheet } from '../utils/atomize'

export const getEngine = <R extends TTyped.Shape>() => untypedEngine as TTyped.TypedEngine<R>

const untypedEngine = {
    WEB: $web,
    NATIVE: $native,
    HOT: $hot,
    IF: $if,
    IFELSE: $ifelse,
    WIDTH: $width,
    THEMED: p => p,
    STYLE: (...p) => p,
    $atomizeSheet: atomizeSheet as any,
    $mergeSheets: mergeSheets as any,
    $atomizeRuleset: atomizeRuleset as any,
    ATOMIZE: (...pars: any[]) => atomizeRuleset(pars, null, '$atomize'),
    $mergeRulesets: mergeRulesets as any,
    $toClassNames: toClassNamesWithQuery as any
} as TTyped.TypedEngine<TTyped.Shape>
