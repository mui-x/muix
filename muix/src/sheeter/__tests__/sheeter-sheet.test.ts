import { $W, $T, $V, $I, V, T, I, } from 'reactxx-typings'
import { getEngine } from '../utils/get-engine'

//import {  } from "reactxx-sheeter"
import { initPlatform, Shape as ShapeLow, theme } from "./init-platform.t"

interface Shape extends ShapeLow {
  sheetQuery: {color: string}  
}

const {
  $themed, $if, $web, $native, $rules,
  $atomizeSheet, $mergeSheets, $atomizeRuleset
} = getEngine<Shape>()

describe("SHEET", () => {
  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let sheet

    it("01 null", () => {
      const sheet = $atomizeSheet(null)
      expect(sheet).toMatchSnapshot()
    })
    it("02 {}", () => {
      const sheet = $atomizeSheet({})
      expect(sheet).toMatchSnapshot()
    })
    it("03 backgroundColor: 'red'", () => {
      const sheet = $atomizeSheet({
        root: $rules<V>({ backgroundColor: 'red' })
      })
      expect(sheet).toMatchSnapshot()
    })
    it("04 merge", () => {
      const sheet = $mergeSheets(null)
      expect(sheet).toMatchSnapshot()
    })
    it("05 merge", () => {
      const sheet = $mergeSheets([])
      expect(sheet).toMatchSnapshot()
    })
    it("06 merge", () => {
      const sheet = $mergeSheets([null, {}, undefined])
      expect(sheet).toMatchSnapshot()
    })
    it("07 merge", () => {
      const sheet = $mergeSheets([
        null, {}, undefined,
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'green' }),
          label: $rules<T>({ color: 'green' }),
        }) as any,
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("08 merge", () => {
      sheet = $mergeSheets([
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'red' }),
          label: $rules<T>({ color: 'red' }),
        }),
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'green' }),
          webOnly: $rules<$W>({ color: 'green' }),
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("09 merge", () => {
      sheet = $mergeSheets([
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'red' }),
        }),
        $atomizeSheet({
          root: $rules<V>($web<V>({ color: 'green' })),
        }),
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'blue' }),
        }),
      ])
      expect(sheet).toMatchSnapshot()
    })
    it("10 with theme", () => {
      sheet = $atomizeSheet(
        $themed(theme => ({
          root: $rules<V>({
            backgroundColor: theme.primary.normal.backgroundColor
          }),
          label: $rules<T>(
            theme.primary.normal,
            { margin: 10 },
            $if<V>(null)
          ),
          webOnly: $rules<$W>(
            $atomizeRuleset<$W>([theme.secondary.disabled], null, 'theme.secondary.disabled')
          )
        })), theme)
      expect(sheet).toMatchSnapshot()
    })
    it("11 sheet source mutated", () => {
      const sheetSource: any = {
        root: {
          backgroundColor: 'red',
          $web: { color: 'green' }
        }
      }
      expect(sheetSource).toMatchSnapshot()
      sheet = $atomizeSheet(sheetSource)
      expect(sheetSource).toMatchSnapshot()
      expect(sheet).toMatchSnapshot()
      sheet = $atomizeSheet(sheet)
      expect(sheet).toMatchSnapshot()
    })
    it("12 merge ERROR, sheets must be atomized first", () => {
      const fnc = () => $mergeSheets([
        {
          root: $rules<V>({ backgroundColor: 'red' }),
          label: $rules<T>({ color: 'red' }),
        },
        $atomizeSheet({
          root: $rules<V>({ backgroundColor: 'green' }),
          webOnly: $rules<$W>({ color: 'green' }),
        }),
      ])
      expect(fnc).toThrow(/.*/)
    })
  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
