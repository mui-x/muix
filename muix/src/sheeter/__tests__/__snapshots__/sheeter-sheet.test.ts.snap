// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHEET ## NATIVE ## 01 null 1`] = `null`;

exports[`SHEET ## NATIVE ## 02 {} 1`] = `Object {}`;

exports[`SHEET ## NATIVE ## 03 backgroundColor: 'red' 1`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 04 merge 1`] = `null`;

exports[`SHEET ## NATIVE ## 05 merge 1`] = `null`;

exports[`SHEET ## NATIVE ## 06 merge 1`] = `Object {}`;

exports[`SHEET ## NATIVE ## 07 merge 1`] = `
Object {
  "label": Array [
    Object {
      "color": "green @sheet.label",
    },
  ],
  "root": Array [
    Object {
      "backgroundColor": "green @sheet.root",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 08 merge 1`] = `
Object {
  "label": Array [
    Object {
      "color": "red @sheet.label",
    },
  ],
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root",
    },
    Object {
      "backgroundColor": "green @sheet.root",
    },
  ],
  "webOnly": Array [
    Object {
      "color": "green @sheet.webOnly",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 09 merge 1`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root",
    },
    Object {
      "backgroundColor": "blue @sheet.root",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 10 with theme 1`] = `
Object {
  "label": Array [
    Object {
      "backgroundColor": "darkgreen @sheet.label[0]",
      "color": "white @sheet.label[0]",
    },
    Object {
      "margin": "10 @sheet.label[1]",
    },
  ],
  "root": Array [
    Object {
      "backgroundColor": "darkgreen @sheet.root",
    },
  ],
  "webOnly": Array [
    Object {
      "backgroundColor": "blue @theme.secondary.disabled",
      "color": "darkgray @theme.secondary.disabled",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 1`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red",
    },
    false,
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 2`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root[0]",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 3`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root[0]",
    },
  ],
}
`;

exports[`SHEET ## NATIVE ## 11 sheet source mutated 4`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red @sheet.root[0]",
    },
  ],
}
`;

exports[`SHEET ## WEB ## 01 null 1`] = `null`;

exports[`SHEET ## WEB ## 02 {} 1`] = `Object {}`;

exports[`SHEET ## WEB ## 03 backgroundColor: 'red' 1`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 04 merge 1`] = `null`;

exports[`SHEET ## WEB ## 05 merge 1`] = `null`;

exports[`SHEET ## WEB ## 06 merge 1`] = `Object {}`;

exports[`SHEET ## WEB ## 07 merge 1`] = `
Object {
  "label": Array [
    Array [
      ".b { color:green } @sheet.label",
    ],
  ],
  "root": Array [
    Array [
      ".a { background-color:green } @sheet.root",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 08 merge 1`] = `
Object {
  "label": Array [
    Array [
      ".b { color:red } @sheet.label",
    ],
  ],
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root",
    ],
    Array [
      ".c { background-color:green } @sheet.root",
    ],
  ],
  "webOnly": Array [
    Array [
      ".d { color:green } @sheet.webOnly",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 09 merge 1`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root",
    ],
    Array [
      ".b { color:green } @sheet.root/$web",
    ],
    Array [
      ".c { background-color:blue } @sheet.root",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 10 with theme 1`] = `
Object {
  "label": Array [
    Array [
      ".d { color:white } @sheet.label[0]",
      ".c { background-color:darkgreen } @sheet.label[0]",
    ],
    Array [
      ".e { margin:10px } @sheet.label[1]",
    ],
  ],
  "root": Array [
    Array [
      ".c { background-color:darkgreen } @sheet.root",
    ],
  ],
  "webOnly": Array [
    Array [
      ".a { color:darkgray } @theme.secondary.disabled",
      ".b { background-color:blue } @theme.secondary.disabled",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 1`] = `
Object {
  "root": Array [
    Object {
      "backgroundColor": "red",
    },
    [Function],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 2`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root[0]",
    ],
    Array [
      ".b { color:green } @sheet.root[1]/$web",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 3`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root[0]",
    ],
    Array [
      ".b { color:green } @sheet.root[1]/$web",
    ],
  ],
}
`;

exports[`SHEET ## WEB ## 11 sheet source mutated 4`] = `
Object {
  "root": Array [
    Array [
      ".a { background-color:red } @sheet.root[0]",
    ],
    Array [
      ".b { color:green } @sheet.root[1]/$web",
    ],
  ],
}
`;
